Collapse option not working for the the BO category tree

crezzur opened the pull request:
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 1.7.7.x
| Description?      |  Resolve issue "collapse" option not working for the the BO category tree (old theme) Affected pages: Faceted search, other pages and modules which use the old theme tree version
| Type?             | bug fix 
| Category?         | BO 
| BC breaks?        | 
| Deprecations?     | 
| Fixed ticket?     | Fixes #23558.
| How to test?      | Displays an error when closing a (sub)category folder in the category tree.
| Possible impacts? | category tree in module Faceted search, other pages and modules which uses the old theme category tree.


<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/23625)
<!-- Reviewable:end -->

Commit:
Update tree.js

Resolve issue "collapse" option not working for the the back-office tree (old theme)
Affected pages: Faceted search, other pages and modules which use the old theme tree version

NeOMakinG commented on admin-dev/themes/default/js/tree.js:
```diff
@@ -213,13 +210,13 @@ Tree.prototype =
 					organizeTree();
 					$('#' + idTree).tree('init');
-					that.$element.find("label.tree-toggler").each(
+					this.$element.find("label.tree-toggler").each(
```
What's the reference of `this` there ? It looks like it's scope is the anonymous function

crezzur replied:
This is a mistake it should be $(this)

The issue is that the current code uses variables for $(this) 
`var that = $(this);` and `var thatOne = $(this);`

However these variables are not passed in the functions `function(content)` which causes the error.

NeOMakinG replied:
that wasn't accessible ? It looks like it's in the upperscope, it should be accessible ! strange thing there

NeOMakinG commented:
I asked a question only on one this, but every `$(this)` scope looks strange to me

Commit:
Update tree.js

Typo

sowbiba commented on admin-dev/themes/default/js/tree.js:
```diff
@@ -213,13 +210,13 @@ Tree.prototype =
 					organizeTree();
 					$('#' + idTree).tree('init');
-					that.$element.find("label.tree-toggler").each(
+					$(this).$element.find("label.tree-toggler").each(
```
hard for me to see here what $(this) refers to

NeOMakinG replied:
It looks like it refers to `"label.tree-toggler, .icon-folder-close, .icon-folder-open"` but the actual behavior refers to the tree, I don't know which one is right, I think we need to checkout @crezzur branch in order to review this, overwise you should maybe write a good description to explain it a bit more :)

crezzur replied:
Not sure what the issue is quys? the element $(this) is a jQuery element which holds the data that was clicked.

As you can see the current code already use the element $(this) but puts it into a variable that and thatOne .
In the current version the variable cannot be accessed from the function. Just try it and you will see i am correct.

NeOMakinG reviewed (dismissed):
Tested it and every changed line was using wrong scoped reference, LGTM to me ;)

khouloudbelguith commented:
Hi @crezzur,

Thanks for your fix.
This PR fixes a regression in PS1.7.7, you need to target the `1.7.7.x` branch.

Thank you!

crezzur commented:
> 
> 
> Hi @crezzur,
> 
> Thanks for your fix.
> This PR fixes a regression in PS1.7.7, you need to target the `1.7.7.x` branch.
> 
> Thank you!

Added the 1.7.7.x branch

Progi1984 commented:
@crezzur No your are targeting `develop` branch and not the `1.7.7.x` branch

crezzur commented:
> 
> 
> @crezzur No your are targeting `develop` branch and not the `1.7.7.x` branch

I had added the 1.7.7.x branch at start but it was removed. I will reverse these changes now.
**Changes**
PierreRambaud removed the 1.7.7.x label 5 days ago
prestonBot added the develop label 5 days ago

Progi1984 commented:
@crezzur I think you misunderstood me : your PR targets the `develop` branch :
![image](https://user-images.githubusercontent.com/1533248/111483370-06adc500-8735-11eb-84b8-01847967ffef.png)

crezzur commented:
@Progi1984  No idea how to change it

NeOMakinG commented:
@crezzur I think you need to change the target branch to 1.7.7.x, and then rebase your branch on the 1.7.7.x branch, if the rebase is too hard, we can just open a new PR or cherry pick the commit on a new PR

crezzur commented:
@NeOMakinG Did i do it correct now?

NeOMakinG commented:
> @NeOMakinG Did i do it correct now?

You just changed the base branch, you need to rebase it into 1.7.7.x now

PierreRambaud commented:
The PR has been rebased on the 1.7.7.x branch ;)

khouloudbelguith commented:
Hi @crezzur,

I have an error in the console when we click on the Expand button.
Steps to reproduce the issue:
1. Go to BO > Module Manager page and search for the ps_facetedsearch module
2. Click on configure 
3. Add a new template
4. In the `Categories used for this template:` section > Click on `Expand`
5. See an error in the console 
![image](https://user-images.githubusercontent.com/16067358/115852818-c7a41b00-a428-11eb-94a5-da4ccb37b36f.png)

**Without PR** => ok
https://drive.google.com/file/d/1lgXB93jAf_VQERS-2nEJ-CDZWvyfHZW4/view?usp=sharing

Also when the multistore context is enabled, with this PR, the tree issue is fixed in the `Choose shop association` section
![image](https://user-images.githubusercontent.com/16067358/115853409-70eb1100-a429-11eb-8edc-a0ea7328353e.png)
With PR => OK in different pages (facetedsearch, attributes, features, Customer Group,..) only in the `Add new shop page`.
Steps to reproduce the issue:
1. Go to BO > Shop Parameters > General page and enable multistore
2. Go to BO > Advanced Parameters > Multistore page 
3. Click on "Add new Shop"
4. Click on `Collapse All`
5. Click on `Check All`
6. Click on `Expand All`
7. Try to close a subcategory folder => NOK

https://drive.google.com/file/d/1o_Ew35RLOhKdigh4Ufaq1ketDNAfjgjJ/view?usp=sharing

This issue should be fixed in this PR?

Thank you!

Commit:
Update tree.js

Error pointed out by @khouloudbelguith, 

Expand all button will fire error cannot find id of tree which needs to expand. Error can be found on line 219 of the expandAll function
Error line: $(this).$element.find("label.tree-toggler").each(

Since we already know the tree id from the click we can us the idTree parameter. 
Resolve error: $('#' + idTree).find("label.tree-toggler").each(

The line "$(this).$element" is not needed since we can use the Jquery solution to find by element ID instead of the this function. In this case we use  "$('#' + idTree)" which prevents any errors in the future.

crezzur commented:
> 
> 
> Hi @crezzur,
> 
> I have an error in the console when we click on the Expand button.
> Steps to reproduce the issue:
> 
>     1. Go to BO > Module Manager page and search for the ps_facetedsearch module
> 
>     2. Click on configure
> 
>     3. Add a new template
> 
>     4. In the `Categories used for this template:` section > Click on `Expand`
> 
>     5. See an error in the console
>        ![image](https://user-images.githubusercontent.com/16067358/115852818-c7a41b00-a428-11eb-94a5-da4ccb37b36f.png)
> 
> 
> **Without PR** => ok
> https://drive.google.com/file/d/1lgXB93jAf_VQERS-2nEJ-CDZWvyfHZW4/view?usp=sharing
> 
> Also when the multistore context is enabled, with this PR, the tree issue is fixed in the `Choose shop association` section
> ![image](https://user-images.githubusercontent.com/16067358/115853409-70eb1100-a429-11eb-8edc-a0ea7328353e.png)
> With PR => OK in different pages (facetedsearch, attributes, features, Customer Group,..) only in the `Add new shop page`.
> Steps to reproduce the issue:
> 
>     1. Go to BO > Shop Parameters > General page and enable multistore
> 
>     2. Go to BO > Advanced Parameters > Multistore page
> 
>     3. Click on "Add new Shop"
> 
>     4. Click on `Collapse All`
> 
>     5. Click on `Check All`
> 
>     6. Click on `Expand All`
> 
>     7. Try to close a subcategory folder => NOK
> 
> 
> https://drive.google.com/file/d/1o_Ew35RLOhKdigh4Ufaq1ketDNAfjgjJ/view?usp=sharing
> 
> This issue should be fixed in this PR?
> 
> Thank you!

Issue resolved by adding $('#' + idTree) to select the correct tree to expand on request.

sowbiba commented on admin-dev/themes/default/js/tree.js:
```diff
@@ -213,13 +210,13 @@ Tree.prototype =
+				{ controller:'AdminProducts', token:currentToken, action:'getCategoryTree', type:idTree, fullTree:1, selected:selected, inputName:name, useCheckBox:useCheckBox },
 				function(content)
 				{
 					$('#' + idTree).html(content);
```
can be improved by putting `$('#' + idTree)` in a variable. But it's not blocking. Thanks @crezzur ;)

crezzur replied:
Added a new commit where the `$('#' + idTree)` is stored in the variable called  `targetTree`

Commit:
Update tree.js

Optimization of expandAll function added Jquery selector $('#'+idTree) into targetTree variable since the Jquery selector is used on 7 locations, requested by @sowbiba

sowbiba reviewed (dismissed):
👌🏾

khouloudbelguith commented:
Hi @crezzur,

The first issue is fixed but the second problem still exists.
Steps to reproduce the issue:

1. Go to BO > Shop Parameters > General page 
2. Enable Multistore context
3. Go to BO > Advanced Parameters > Multistore page 
4. Click on "Add new shop"
5. In the "Associated categories" section, click on `Collapse All`
6. Click on `Check All`
7. Click on `Expand All`
8. Try to close a folder
9. See error: Cannot be closed

I attached a screen record
https://drive.google.com/file/d/1XfrDGN6BgR_LmCS8ALnpO0fDeUEVR7VO/view?usp=sharing

Thank you!

crezzur commented:
@khouloudbelguith
The error is showing instantly when you click on the **expand all** button. The reason it is displayed in the **Add shop** treeview only is because this treeview has the **search option** activated.

The error is caused by the function call `organizeTree(); `(tree.js line 218).
Looking into this function it looks like its a dead part of code.

inside the function `organizeTree()` a call to the function `getCategoryById()` is made. Looking into the function `getCategoryById()` the HTML attribute `name="id_parent"` is searched. 

Searching into the whole folder `STORE\BACKFOFFICE\themes\default\` and more specific `STORE\BACKFOFFICE\themes\default\template\helpers\tree\` for the HTML attribute `name="id_parent"` returns negative.

concluding this i am guessing the functions organizeTree() and getCategoryById() are leftovers from something what is already removed. If you come to the same conclusion i will add a pull request to remove those functions.

khouloudbelguith commented:
Hi @crezzur,

Thanks for your feedback.
So this issue of creating a shop will be fixed in another PR?

Thank you!

Commit:
Update tree.js

The error is showing instantly when you click on the expand all button. The reason it is displayed in the Add shop treeview page only is because this treeview has the search option activated.

The error is caused by the function call organizeTree(); (tree.js line 218).
Looking into this function it looks like its a dead part of code.

inside the function organizeTree() a call to the function getCategoryById() is made. Looking into the function getCategoryById() the HTML attribute name="id_parent" is searched.

Searching into the whole folder STORE\BACKFOFFICE\themes\default\ and more specific STORE\BACKFOFFICE\themes\default\template\helpers\tree\ for the HTML attribute name="id_parent" returns negative.

concluding this i am guessing the functions organizeTree() and getCategoryById() are leftovers from something what is already removed.

crezzur commented:
@khouloudbelguith added the changes to this pull request.

khouloudbelguith commented:
Hi @crezzur,

It is ok :heavy_check_mark: 
https://drive.google.com/file/d/1IDkA6nqb1P4OPc22qNTl32c-QXBYcRYD/view?usp=sharing

Thank you!
